### ISSUE TITLE
externs as a constructor parameter to other externs

### ISSUE THREAD

**Issue by cc10512** [p4lang/p4c#1785]
Per the LDWG discussion on 3/4, here is the full example of the code for passing externs as constructor parameters to other externs. Details of the proposal are https://github.com/p4lang/p4-spec/wiki/Externs-as-constructor-parameters and included below for your reference.

# Problem
We encountered several instances in which the state of an extern needs to be manipulated from the control plane. For example hash functions that compute the hash on a field list specified by the control plane.

# Current state
The current language specification requires that all the constructor parameters are compile time constants. While we can ask the user to enumerate all the possible field lists that are available as candidates for computing the hash, because lists and tuples are internal compiler data structures, we have no means of identifying the selected list from the control plane. Moreover, the elements should not capture the values at construction time, but rather use the values that are in the fields when the `get` method is invoked.

# Proposed solution
## Pass an extern as a constructor parameter to another extern
By passing an extern as a constructor parameter to another extern we can:
  - use the name of the extern instance or constructor parameter in the control plane to identify a particular encapsulate state, in our case a field list
  - since the semantics of the extern is entirely decided by the compiler, we decide what should happen to the state of the argument passed as a constructor parameter. In our example, the extern instance is passed as a reference and values are captured only when we invoke the method that needs them. Think of the same semantics as a match key for a table.

### Issues
What happens if the same instance is passed to multiple externs? We can let the programmers to express what should be the semantics using parameter labels. `in` will continue to act as a read-only, copy-in parameter. No direction or `ref` can specify a parameter that is acted upon by the control plane and is called by reference.
### Code changes
Trivial -- require removing 4 lines of code that check parameters

# Example

```
#include <core.p4>

header hdr1 {
    bit<8> f1;
    bit<8> f2;
    bit<16> res;
}

header hdr2 {
    bit<8> f3;
    bit<8> f4;
    bit<8> f5;
    bit<8> f6;
}

@metadata @name("standard_metadata")
struct standard_metadata_t {
    bit<9>  ingress_port;
    bit<9>  egress_spec;
    bit<9>  egress_port;
    bit<1>  drop;
}

struct hdrs {
    hdr1 h1;
    hdr2 h2;
}

/// Field list extern -- identifies a set of fields
extern FieldList<W> {
   FieldList(W field_list);
};

extern FieldLists<T> {
   FieldLists(T fls);
};

enum HashAlgorithm_t {
  IDENTITY,
  CRC32,
  CRC32_CUSTOM,
  CRC16,
  CRC16_CUSTOM,
  ONES_COMPLEMENT16,  /// One's complement 16-bit sum used for IPv4 headers,
                      /// TCP, and UDP.
  TARGET_DEFAULT      /// target implementation defined
}

/// An extern that requires the control plane to choose which field list is
/// participating in the hash computation
extern DynamicHash<W> {
    DynamicHash(HashAlgorithm_t algo, FieldLists<_> fls);
    /// the control plane needs to select the FieldList.
    W get();
    W get(in W base, in W max);

    /// LDWG discussion proposed to pass the field list to the get method
    /// However, again, the goal here is for the control plane to set
    /// which one of the lists is desired, so it is not clear to me
    /// why get with the whole set is more intuitive.
    /// W get(FieldLists<_> fls);

    /// the data plane may want to set a default
    void setDefault(FieldList<_> fl);
}

control c(inout hdrs h, inout standard_metadata_t std_meta) {
    /// Declare several filed lists
    FieldList<tuple<bit<8>, bit<8>, bit<8>>>({ h.h1.f1, h.h1.f2, h.h2.f3 }) fl0;
    FieldList<tuple<bit<8>, bit<8>, bit<8>, bit<8>>>({ h.h1.f1, h.h2.f4, h.h2.f5, h.h2.f6 }) fl1;

    /// construct a dynamic hash extern
    DynamicHash<bit<16>>(HashAlgorithm_t.CRC16, FieldLists<_>({fl0, fl1})) hash0;
    action a () {
        h.h1.res = hash0.get();  // correct.
    }
    action dropIt() {
        std_meta.drop = 1;
    }

    table t {
        key = {
            h.h1.f1: exact;
            h.h2.f3: exact;
        }
        actions = {
            a;
            dropIt;
        }
        default_action = dropIt;
    }

    apply {
        t.apply();
    }
}

control proto(inout hdrs h, inout standard_metadata_t s);
package top(proto p);

top(c()) main;
```

**Comment by mbudiu-vmw** [comment]
I think that what you want can be done simply with:

```
extern T hash<T, T1, T2>(in T1 list1, in T2 list2);
extern T hash<T, T1, T2, T3>(in T1 list1, in T2 list2, in T3 list3);
...
```
The you just call the function you want:
```
bit<16> x = hash({ f1, f2, f3}, { f4, f5, f6}, { f7, f8 });
```

**Comment by cc10512** [comment]
@mbudiu-vmw see my comment:
```
    /// LDWG discussion proposed to pass the field list to the get method
    /// However, again, the goal here is for the control plane to set
    /// which one of the lists is desired, so it is not clear to me
    /// why get with the whole set is more intuitive.
    /// W get(FieldLists<_> fls);
```

**Comment by mbudiu-vmw** [comment]
First, it's very simple. 
Second, it does not involve any changes to the language.
Third, I think it does what you need.

**Comment by hanw** [comment]
I think one of the design issue is the P4-16 language features: anonymous instantiation of externs and struct initializer made control plane API generation difficult, because the lack of names associated with the thing that is being created.

**Comment by hanw** [comment]
The point of the proposal is not about how to express computation (hash) in dataplane, it is about how to more easily extracting info for control plane API.

**Comment by mbudiu-vmw** [comment]
there is no anonymous instantiation. This was a function call, but we can also use an extern with a single method. Then you have the instance name available in the control-plane.

**Comment by hanw** [comment]
there is no anonymous instantation in this example, but we have come across other similar issues with anonymous parser instantiation.

**Comment by mbudiu-vmw** [comment]
what is anonymous instantiation?

**Comment by mbudiu-vmw** [comment]
parsers are inlined

**Comment by hanw** [comment]
Switch(Parser(), Control(), Deparser()) main. I meant the `Parser()`.

**Comment by cc10512** [comment]
Unfortunately it does not. 
First, it is not clear how the control plane can identify and select a list. The set of lists available for the control plane should be fixed. Having a function instantiating various lists is not a fixed set.
Second, the call to hash with the full set is confusing: the hash will be computed on only one of the lists. Which one it is not clear. I believe that it is easier to explain that the extern maintains state that is set by the control plane and the `get()` method operates on that state.

**Comment by mbudiu-vmw** [comment]
I don't think it's any harder to explain, the signatures are very similar.
The control-plane will just select an index between 0 and 2 - the index of the list.
```
extern MuxHash<T> {
    MuxHash();
    extern T get<T1, T2>(in T1 list1, in T2 list2);
    extern T get<T1, T2, T3>(in T1 list1, in T2 list2, in T3 list3);
}

MuxHash<bit<16>> mh();
mh.get( { f1, f2 }, { f3, f4 } );
```

The control-plane gets to see the name mh. The control plane API to mh is mh.listIndex(int index);

**Comment by mbudiu-vmw** [comment]
@hanw: these are not anonymous, they are bound to parameter names for the Switch package.

**Comment by jafingerhut** [comment]
I am guessing you (Calin, Han) consider there to be significant advantages for ongoing maintenance of the P4 program and co-developed control plane software if there is a _name_ for each of the possible field lists?

Certainly I can imagine scenarios where using an integer index into a list as Mihai suggests would have a significant disadvantage there.  For example, imagine a scenario where you have 4 field lists for a single `get` call, and the system developers decide they no longer want to support the 3rd one, then afterwards the one that used to be index 3 becomes index 2 with the new program.  If there were names associated with each one, the removed name would easily be detected as no longer valid, vs. silently aliased to a different field list.

If that is a significant factor for the design here, then as an off-the-top-of-my-head suggestion (sorry, no deep thought put into it yet), perhaps something like Mihai's suggested syntax, but with names annotating each field list, is worth a moment's consideration?  For example:

```
MuxHash<bit<16>> mh();
mh.get( @name("fl_A") { f1, f2 },
        @name("fl_B") { f3, f4 } );
```
That particular syntax might not be good for generating the control plane API names.  The intent of this comment is primarily to bring out the desired properties of the design, not only in the data plane but also the control plane, to help focus any future suggestions.

**Comment by mbudiu-vmw** [comment]
We don't support annotations on expressions, but we could perhaps put the names in the constructor as strings.

**Comment by mbudiu-vmw** [comment]
Or we could have a `@name` annotation on the instance declaration.

**Comment by jafingerhut** [comment]
Mihai, if you mean a `@name` annotation on the declaration of instance `mh` from your earlier comment, a single such annotation is insufficient to name each field list independently, isn't it?  I could be missing the meaning of your suggestion, though.

**Comment by cc10512** [comment]
Thank you @jafingerhut for articulating our problem so crisply.
Why should we use annotations when the solution we proposed to directly name the instances?

**Comment by mbudiu-vmw** [comment]
An annotation can have any number of expressions. Probably we would call it something else, though.

**Comment by mbudiu-vmw** [comment]
```
MuxHash mh("list1", "list2", "list3");

Or

@lists("list1", "list2". "list3")
MuxHash mh();
```

**Comment by cc10512** [comment]
Sorry, I'm confused! What exactly are we trying to optimize? The usability of the language or The amount of work that we need to do in the compiler? In either case, the change we're asking for is compatible with previous programs and the work is to elide a check. And it would make for an easier use of the APIs and maintenance.

If there is anything that will break, I'm happy to consider it those instances. Otherwise, please trust that we have looked at all these options before coming up with the proposal.

**Comment by jafingerhut** [comment]
I do not want to put words into Mihai's mouth, here, although I may get lucky and express his intent.

Your original code snippets have a similar property similar to the reason that p4c currently doesn't always do recirculate/resubmit/clone with field lists the same way a P4_14 program does: they name fields in one part of the code, but access the values of those fields in a completely different part of the execution flow.  The place where `get()` is called in your code examples do not locally mention _right there_ in that call what field values it can read.  No other P4_16 extern calls can do that right now, apparently by explicitly design choice of the authors of the P4_16 language spec.  Read through Section 6.7.1 "Justification" of the P4_16 language spec, and if you can do that, and explain to me how your proposed `get()` call reads the value of fields in the field list _right there at the time the get() call occurs in the execution flow_, then I will have learned something new.

Basically, it seems to me that introducing such a capability into P4_16 makes some compiler writers a bit nervous about how to precisely nail down where the "mentions" of the names of values appear in the source code, vs. where the actual run-time read/write of that value occurs.  For P4_14 it seems to me that this has been decided quite a while ago, but only documented in the P4_14 spec to somewhat greater degrees of precision as new P4_14 spec releases have occurred.

**Comment by mbudiu-vmw** [comment]
Adding references is a major change to the language which i believe will also decrease its usability and make it much harder to implement on many targets. People will be able to write many programs that may be very hard to support.

**Comment by jafingerhut** [comment]
Note: My previous comment is not intended to say "I think there should never be a way to write a P4_16 program following your proposal".  It is intended to say "I cannot see any way to write an extern that obeys the P4_16 v1.1.0 specification, and does what you want it to, at the same time."

P4_16 v1.x.0 for x >= 2?  Sure, that may be possible, if people want to enhance the language spec in appropriate ways.

**Comment by jafingerhut** [comment]
If I can try to summarize what I think the P4_16 v1.1 language spec says about the code example given in the original comment:

All constructor calls are executed once, at compile time.  They are not executed once for each processed packet.

The call to `hash0.get()` is executed at run-time for each packet that invokes action `a`.  What information about the packet does it have access to?  _None_, because it has an empty parameter list.  Or more generally, it could have access to data about the packet, but only if an earlier call to a method of object `hash0` passed those values about the packet to that method.  If `hash0.get()` is the first method call on extern `hash0` while processing a packet, the extern has never learned anything about that packet before, so it knows nothing about it.

**Comment by jafingerhut** [comment]
As a fairly minor variation on the P4 code example in the original comment, that does provide names (via action names) to each individual field list, that are stable across program maintenance, and does not use any P4_16 language semantics outside of what is in the spec already, consider this:
```
    action hash_using_fl0 () {
        h.h1.res = hash0.get({ h.h1.f1, h.h1.f2, h.h2.f3 });
    }
    action hash_using_fl1 () {
        h.h1.res = hash0.get({ h.h1.f1, h.h2.f4, h.h2.f5, h.h2.f6 });
    }
    action dropIt() {
        std_meta.drop = 1;
    }

    table t {
        key = {
            h.h1.f1: exact;
            h.h2.f3: exact;
        }
        actions = {
            hash_using_fl0;
            hash_using_fl1;
            dropIt;
        }
        default_action = dropIt;
    }
```

**Comment by cc10512** [comment]
This is actually the behavior we'd like to capture: have a single program point that defines all the available variants (the constructor) and then multiple runtime calls that extract whatever values are there in the headers at runtime. It is the responsibility of the control plane to ensure that those values are valid, and the data plane programmer can ensure that at least all headers are valid by invoking the action in a table that verifies it. Indeed it is a bit round-about, but not different from any other header validity check. I guess your concern is how is this playing with our desiderata of allowing only safe programs, and indeed, there is a potential gap because the compiler relinquishes the control to the programmer. I don't think there is an easy way to check for this. However, I do not think this is any different than having the call list all the possible combinations, because, I _can_ write an analysis that checks that everything is defined at every call site, just as well as explicitly listing all parameters at each call site. Am I missing something here?

And finally, externs are already passed by reference, so there is no real difference between this proposal and the existing language definition. I think the `ref` direction is actually not needed.

**Comment by hanw** [comment]
One way to think about the example is to realize that maybe not all externs are pure-dataplane objects. Some externs are control-plane-managed, and dataplane-executed objects. For example, the dynamic hash extern in this example, AND, the table abstraction (Even though in P4, we choose to treat table specially).

```
extern DynamicHash<W, T> {
   DynamicHash(FieldLists<_> fls)
   W get();                // API invoked by dataplane.
   W get(T t);
}

// pseudo code for the get() method implementation.
W m.get() {
   FieldList fl = m.getCurrentFieldList(m.fieldLists);
   return m.hash.get(fl.fields);
}
```

Similarly, here the pseudo code for table.apply()
```
apply_result(m) m.apply() {
    ...
    var lookupKey = m.buildKey(m.key); // using key block
    action RA = m.table.lookup(lookupKey);
    ...
}
```

We never say `table.apply()` does not have access to the key it tries to match on.

**Comment by jafingerhut** [comment]
Han, if you can read Section 6.7.1 "Justification" of the P4_16 language spec, and this comment: https://github.com/p4lang/p4c/issues/1785#issuecomment-471792231

and _then_ explain to me how your proposed `get()` call can access any contents of the packet at all, without modifying the P4_16 language spec, I would learn something.

**Comment by jafingerhut** [comment]
As far as I can tell right now, these lines of the example code in the original comment:
```
    /// Declare several filed lists
    FieldList<tuple<bit<8>, bit<8>, bit<8>>>({ h.h1.f1, h.h1.f2, h.h2.f3 }) fl0;
    FieldList<tuple<bit<8>, bit<8>, bit<8>, bit<8>>>({ h.h1.f1, h.h2.f4, h.h2.f5, h.h2.f6 }) fl1;
```
are violations of the parts of the current P4_16 language spec that extern constructor calls are restricted to have only compile-time constants as arguments.

Again, I bring up these issues not because I like to language-lawyer things out of existence.  I bring them up to point out that if you want the proposed mechanism, as described, it seems like the best way to go about it is to change the language spec to define what the syntax in your proposal means.

**Comment by jafingerhut** [comment]
Han, I agree that `<table_name>.apply();` is special in P4_16, compared to extern method / function calls.

It is _explicitly defined in the spec_, and understood by compiler writers, that it _will_ read all fields of the table search key, and then perform exactly one of the actions, including any reading or writing of fields in that action's body.

P4_16 v1.1 extern functions and methods are explicitly defined in the language spec to not be able to access any variable in a P4 program at all, except via copy-in/copy-out semantics of its arguments.

**Comment by cc10512** [comment]
@jafingerhut The proposal was to change the language spec. I brought it up in the LDWG meeting (https://github.com/p4lang/p4-spec/wiki/Externs-as-constructor-parameters) and I was explicitly asked to open an issue so that people can comment. I will move the issue to p4-spec if that's more clear.
